### PR TITLE
Add groupMapReduce method to IterableOnceOps

### LINKED ISSUE #8: Add groupMapReduce to Iterator
Currently `groupMapReduce` is defined in `IterableOps` so it's not available for `Iterator`, but it easily could be because it doesn't require building an inner collection of the collection type, unlike `groupBy` and `groupMap`. Thus `groupMapReduce` would be hoisted to `IterableOnceOps` instead and be a consuming operation on the iterator.

I suppose the current workaround would be to use `to(LazyList)` on an iterator and use `groupMapReduce` on that.

### PR DESCRIPTION
**Pull request by BalmungSan**
Fixes #8 

I hope I correctly followed the structure and naming conventions that you want to enforce in this repo. I based my work here on what Seth is doing in #17 and I am aware that is also the first addition and that the guidelines may change, I subscribed to that PR to follow the discussion and conclusions _(for example here I am following Sébastien advice on the naming of the implicit class)_, but if I miss something please let me know and I will happily push the requested changes.

Any kind of feedback is greatly welcome, but I am personally very interested in three things:
1. Should the implicit class also be an `AnyVal`? I noticed Seth's one isn't and nobody discussed that. Is this just an oversight? Or personal preference? AFAIK for extension methods `AnyVal` should guarantee no instantiation and shouldn't have any funny side-effect since the class is not expected to be used as a type. But maybe I am missing something, maybe it has some problems with binary compatibility or something. Please let me know what you guys think about this.
2. I basically copied the implementation from the one in [**IterableOps**](https://github.com/scala/scala/blob/v2.13.3/src/library/scala/collection/Iterable.scala#L617), with some sightly style changes... but I am worried those changes _(especially using `updateWith`)_ could have a negative impact on performance.
3. I am not sure if that test is enough but being honest I do not have any other ideas. Also, again I tried to copy what was on the stdlib, and [this](https://github.com/scala/scala/blob/2.13.x/test/junit/scala/collection/IterableTest.scala#L58) was the only test I found; but I won't say I read all the code, maybe I missed something?

Also, I would also like to propose adding `groupByTo(factory)(key)` and `groupMapTo(factory)(key)(f)` methods as well. What do you guys think? May I also add them in this PR or in another one? Or don't you find them useful _(I do)_?

### CODE REVIEW

**Comment by sinanspd** [comment]
Is there a specific reason why we would introduce this functionality through a implicit class rather than putting it directly in the Iterator ?

**Comment by NthPortal** [comment]
> Should the implicit class also be an `AnyVal`?

yes

> I noticed Seth's one isn't and nobody discussed that. Is this just an oversight?

possibly an oversight on Seth's part. I noticed it and planned to fix it or comment on it, but was waiting on a more concrete decision wrt package structure (see #24).

> but I am worried those changes (_especially using `updateWith`_) could have a negative impact on performance.

I don't know if it will, but I'd probably just copy the implementation from the stdlib.

> Also, I would also like to propose adding `groupByTo(factory)(key)` and `groupMapTo(factory)(key)(f)` methods as well.

sounds potentially valuable. I would recommend opening a separate ticket for that.

File: src/main/scala/scala/collection/IteratorExtensions.scala
Comment (NthPortal) [inline src/main/scala/scala/collection/IteratorExtensions.scala:16]: any reason this is an extension to `Iterator` and not to `IterableOnceOps`?

Comment (julienrf) [reply src/main/scala/scala/collection/IteratorExtensions.scala:16]: `Iterable` also inherits `IterableOnceOps`, but it already has `groupMapReduce`. Do we have another type than `Iterator` that would benefit from the operation?

Comment (BalmungSan) [reply src/main/scala/scala/collection/IteratorExtensions.scala:16]: I was about to ask the same, what would be the benefit of adding this method to any **IterableOnce**?

Comment (NthPortal) [reply src/main/scala/scala/collection/IteratorExtensions.scala:16]: theoretically someone else can add their own `IterableOnceOps` implementation and might want it

Comment (BalmungSan) [reply src/main/scala/scala/collection/IteratorExtensions.scala:16]: @NthPortal ok that is a valid argument, but one question, why `IterableOnceOps`? Shouldn't it be `IterableOnce`?

Comment (NthPortal) [reply src/main/scala/scala/collection/IteratorExtensions.scala]: the other reason for putting it in `IterableOnceOps` is that that will be the eventual location in 3.1, and there won't be two copies of the method

--------

> Why `IterableOnceOps`? Shouldn't it be `IterableOnce`

no. @julienrf can probably answer this better than I can, but essentially, `IterableOnce` is meant to be barebones, and not contain all of the transformations

File: src/main/scala/scala/collection/next/IterableOnceExtensions.scala
Comment (NthPortal) [inline src/main/scala/scala/collection/next/IterableOnceExtensions.scala:17]: while this particular extension method doesn't require the `CC` or `C` parameters, future extension methods might, so I think it's worth including them for future compatibility

**Review by NthPortal — COMMENTED** [review]
LGTM, excepting the one parameter change I noted.

I do want to make sure this is formally accepted before we merge though. It's not clear to me if the issue is considered an enhancement or bugfix, and if the former, it needs formal acceptance into the scala library

File: src/main/scala/scala/collection/next/IterableOnceExtensions.scala
Comment (BalmungSan) [reply src/main/scala/scala/collection/next/IterableOnceExtensions.scala:17]: You are right, my mistake, I just used **Any** to make it compile and run the tests and forgot to change it 😅

**Comment by BalmungSan** [comment]
> I do want to make sure this is formally accepted before we merge though.

Absolutely, as I said in #24 I believe here should only exist things that are already accepted to be part of a future **Scala** release.
I know that I made a risky move by implementing this before it was formally expressed it will be accepted, but I wanted to give it traction to this initiative.

I believe it would be worth adding a section in the **README** about when it is clear a change is to be accepted. I would propose that a PR MUST always be linked to a previous issue and that the issue MUST be labelled `ACCEPTED` or something like that.

**Comment by NthPortal** [comment]
> I know that I made a risky move by implementing this before it was formally expressed it will be accepted, but I wanted to give it traction to this initiative.

Personally, I only think it is risky if you consider it to have been a significant amount of work. I agree that having an actual proposed/working implementation adds traction to it.

> I would propose that a PR MUST always be linked to a previous issue and that the issue MUST be labelled ACCEPTED; or something like that.

I think we should narrow this slightly to things labeled `enhancement`, but on the whole I agree. I don't think things that are considered bugfixes or similar (e.g. the `batchingGlobal` `ExecutionContext`) that for technical reasons cannot be binary compatible should need formal acceptance though; I think their nature makes them "already accepted" so to speak.

That's actually why I commented about being unsure if this was considered an enhancement or bugfix.

**Comment by BalmungSan** [comment]
I just pushed the _"parameter change"_ requested by @NthPortal as well as modifying the code so it follows the current most voted alternatives of the #24 & #46 proposals.

File: src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala
Comment (NthPortal) [inline src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: I think we should use the exact implementation from the standard library—both for consistency, and to make it easier to apply changes from one to the other if needed

Comment (BalmungSan) [reply src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: Well, the implementation is largely the same on from the stdlib with some minor aesthetical changes which IMH, make the code easier to read.

But, I really do not have strong reasons for not changing them, if any other person believe I should make the change I would do it without problems.

Comment (NthPortal) [reply src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: I think the stdlib likely doesn't use `foreach` for performance reasons, but I'm not sure. I'd rather the PR to make the code cleaner go to the stdlib first though, and then this library can follow suit, if that makes sense.

This is the only thing holding up this PR, to me

Comment (BalmungSan) [reply src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: @NthPortal AFAIK what is holding on this PR is #52 instead of this change or the structure proposals.

About this particular change, as I said I would like to hear what others think.

Comment (NthPortal) [reply src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: I think we can merge this based on https://gitter.im/scala/contributors?at=60523e1d33346c16275ae3af —I think this PR is fairly uncontroversial. it's just moving a method up the collections hierarchy really

Comment (BalmungSan) [reply src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: @NthPortal so should I just push the revert of the style changes and this would be merged?
@SethTisue are you also ok with that _(both the merge and the code style change)_?

Comment (BalmungSan) [reply src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: @NthPortal btw I just realized this:

>I think the stdlib likely doesn't use foreach for performance reasons

Note that the stdlib does uses `foreach` just that using the `for` **sugar syntax**.
My change was just calling that method directly and using `updateWith` instead of `get` + **patten matching** + `put`
The performance should be the same, I just believe the code is cleaner now.

Comment (NthPortal) [reply src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala:33]: my mistake. I thought the stdlib used a while loop for some reason

**Review by NthPortal — APPROVED** [review]

**Comment by NthPortal** [comment]
I've merged this, because I think it's fairly uncontroversial. finalizing the style can come before a release

**Pull request merged**
